### ISSUE TITLE
THEORY: Obtain a rigorous upper bound on the order of field needed to detect linear magma implications

### ISSUE THREAD

**Issue by teorth** [teorth/equational_theories#418]
Relates to #416.  In theory, the Lefschetz principle tells us that if one of our implications is disproven by a linear magma, then this counterexample should be visible in a small finite field.  However, we do not have an actual upper bound on how small this field has to be.  As such, our numerical runs (which have searched cyclic groups up to Z/200Z) do not yet guarantee that we have exhausted *all* linear anti-implications.   Is there a way to use theoretical tools (e.g., Weil conjectures, theory of resultants, etc.) to get a useful upper bound?  Are the bounds better if we permit finite fields of prime power order?  

Numerically, the largest field that generated a new anti-implication was F_{43}.

**Comment by carlini** [comment]
To clarify, 43 was the number that had a new implication on top of all the proofs we've done already. It's possible some value >43 would have given a new proof on top of the other linear equations, but we'd proven it a different way.

It will take a bit longer to check over all equations, but I'll start a run for that to see what it gives.

**Comment by teorth** [comment]
By the way, how difficult would it be to check your code over finite fields rather than over cyclic groups?  I suspect (due to the Chinese remainder theorem) that the only cyclic groups that are actually useful are the ones of prime power order, but also suspect that finite fields of prime power order will be slightly more able to generate anti-implications than their cyclic group counterparts (the two finite rings are similar, but not quite isomorphic, except when the power of the prime is exactly one).

**Comment by carlini** [comment]
Yeah it already does check over only prime powers, I haven't checked what happens when it's not yet.

Okay here's a table that shows:
- For each field size Z/kZ
- The total number of implications that can be refuted at this size
- The number of "new" implications that can't be found at smaller sizes, but were found at this size
- The number of "unique" implications that are found *only* at this size, and not at larger or smaller sizes

|Z/kZ|Total Implications|New Implications|Unique Implications|
|---|--|--|--|
| 2 | 11871871 | 11871871 | 3648439
| 3 | 6346600 | 650536 | 177079
| 4 | 5955741 | 394771 | 132888
| 5 | 4616960 | 216922 | 20366
| 7 | 4778507 | 186150 | 151324
| 8 | 1951389 | 7284 | 532
| 9 | 3679257 | 19552 | 6504
| 11 | 2027957 | 1386 | 32
| 13 | 2579296 | 406 | 0
| 16 | 2591391 | 122 | 28
| 17 | 2030121 | 0 | 0
| 19 | 1253927 | 0 | 0
| 23 | 890125 | 0 | 0
| 25 | 2787114 | 0 | 0
| 27 | 1893552 | 350 | 330
| 29 | 709605 | 0 | 0
| 31 | 610889 | 0 | 0
| 32 | 1030334 | 0 | 0
| 37 | 605101 | 0 | 0
| 41 | 471193 | 0 | 0
| 43 | 544823 | 0 | 0
| 47 | 405577 | 0 | 0
| 49 | 905809 | 134 | 134
| 53 | 302901 | 0 | 0
| 59 | 368179 | 0 | 0


I've only completed the brute force up to 59, so that's why it stops there.

("why is there nothing new at 43?" This is because of the way that I minimized my set of equations. I tried to do a crude approximation of the minimum set cover problem. Basically: given some number of (p,a,b) assignments that refute various implications, pick the smallest number of assignments that refute all the same implications. I do this greedily and pick the largest set that finds new refutations, and choosing a (43,*,*) one was more efficient than taking some smaller values, even though this data shows we could have done the same thing without it.)

**Comment by teorth** [comment]
Just to clarify: are you working with the finite field F_{p^j} or the cyclic group Z/p^j Z?  These are slightly different objects, for instance the latter contains zero divisors.  I'm not sure how finite fields are implemented in your software library but one can for instance work with the polynomial ring F_p[t] quotient out by a degree j irreducible polynomial (the precise choice of polynomial does not matter).

**Comment by carlini** [comment]
Oh I think I understand now. (Sorry, you're quickly approaching the limit of my decade-old undergraduate-math knowledge...)

I am working only over the cyclic group Z/p^jZ. My math library is just the c programming language, which is to say, the math library is practically non-existent. So I don't have any way to handle more complicated objects that I think would be needed for j>1 fields.

Let me explain what I'm doing in more detail and hopefully this matches with my answer. Right now all I do is take each of the equations (e..g, x ◇ y = y ◇ x) and convert it to some code:

```python
def check_eq_43(x, y, p, a, b): 
    return (a*x + b*y)%p == (a*y + b*z)%p
```

I then loop over all triples (p,a<p,b<p). For each of these, I find which equations are satisfied by enumerating over integers x<p,y<p (in general, z,w,...too) and see if the equations is satisfied for all (x,y,...). This gives me the set {satisfies}, {not satisfies} which give potential refutations.

The field question is definitely interesting, but not something what I have right now is ready to handle.

**Comment by teorth** [comment]
Oh, okay.  Probably it doesn't matter - theory (e.g., the Chebotarev density theorem) tells us that an equation that is satisfied over a finite field F_{p^j}, will eventually be satisfied over F_q for some other prime q (possibly significantly larger than p), so in the long run the numerics will exhaust the possibilities in either case.

**Comment by carlini** [comment]
As an update, the first time you can find new counterexamples past 49 is at Z/81Z. (Of course, not a new one for the project.) It can show that
- Equation3102[x==(((y ◇ x) ◇ x) ◇ x) ◇ x] =/> Equation2936[((y ◇ (y ◇ x)) ◇ x) ◇ x] by (k=81,a=3,b=79)
- Equation412[x = x ◇ (x ◇ (x ◇ (x ◇ y)))] =/> Equation617[x = x ◇ (x ◇ ((x ◇ y) ◇ y))] by (k=81,a=4,b=78)

I guess it makes sense they're duals. (It finds the ~first (a,b) that refutes, so that's why it doesn't find (3,79) and (79,3), but that does work too.)

**Comment by AlyciaBHZ** [comment]
We can give a complete answer for the linear fragment over prime fields.

**Setup.** For a prime $p$, interpret the magma operation as $x \diamond y = ax + by$ with $a, b \in \mathbb{F}_p$. Every order-four ETP equation $E: s = t$ becomes a system of polynomial vanishing conditions on $(a,b)$: specifically, $E$ holds for all inputs iff every coefficient difference $c_i(s; a,b) - c_i(t; a,b)$ vanishes, where $c_i \in \mathbb{Z}[a,b]$ with degree $\leq$ depth of the term tree (at most 4 for this family).

This reduces the finite-model search to a scheme-theoretic incidence problem in $\mathbb{A}^2$: each equation cuts out a closed subscheme $X_E = \mathrm{Spec}\, \mathbb{Z}[a,b]/I_E$, and an anti-implication $E_i \not\Rightarrow E_j$ over $\mathbb{F}_p$ is a rational point in the constructible difference $X_{E_i} \setminus X_{E_j}$.

**Content bound.** All coefficient-difference polynomials have integer coefficients with $|\text{coeff}| \leq 2$, so the only prime at which a nonzero defining polynomial can collapse to the zero polynomial is $p = 2$. For odd primes the scheme structure is identical; only the rational-point distribution varies.

**Results (prime fields, exhaustive computation + Groebner certificates):**

For the 810 laws with $\leq 2$ variables:

| $p$ | new anti-implications | cumulative |
|---:|---:|---:|
| 2 | 439,263 | 439,263 |
| 3 | 90,422 | 529,685 |
| 5 | 56,259 | 585,944 |
| 7 | 20,008 | 605,952 |
| 11 | 2,612 | 608,564 |
| 13 | 384 | 608,948 |
| 17 | 0 | 608,948 |
| ... | 0 | 608,948 |

The 47,152 ordered pairs not separated by any prime $\leq 31$ were certified as radical containments $I_{E_j} \subseteq \sqrt{I_{E_i}}$ in $\mathbb{Q}[a,b]$ via exact Groebner bases (Buchberger over $\mathbb{Q}$, exponent bound 16, maximum certificate exponent actually used: 4). **Zero open pairs.** So the anti-implication relation saturates at $p \leq 13$ for all primes, not just tested ones.

For the 2,090 laws with exactly 3 variables, the same method certifies saturation at $p \leq 7$ (1,675,767 unseparated pairs, all certified, 0 open).

**Why this works.** The Bezout bound gives intersection multiplicity $\leq 16$ for independent pairs of degree-4 curves. But the real leverage is that radical containment $I_j \subseteq \sqrt{I_i}$ is decidable by Groebner bases and is a statement about all fields simultaneously, not just tested primes. Once you have the certificate, no further prime scanning is needed.

**To answer the original question directly:** for linear magmas over prime fields $\mathbb{F}_p$, the rigorous upper bound is $p \leq 13$ (2-variable laws) and $p \leq 7$ (3-variable laws). Every anti-implication visible over any prime is already visible by $p = 13$.

---

We'd like to bring this to your attention from our work on the [Omega project](https://github.com/the-omega-institute/automath), which studies the mathematical structures forced by the golden mean shift ($x^2 = x + 1$). The connection to linear magmas is natural: the characteristic equation forces polynomial identities on finite fields, and the coefficient ideal framework above is how we analyze them.

The full writeup is in [Section: Finite-Field Equational Saturation of Linear Magmas](https://github.com/the-omega-institute/automath/blob/dev/theory/2026_golden_ratio_driven_scan_projection_generation_recursive_emergence/sections/body/finite_field_equational_saturation/sec__finite-field-equational-saturation.tex) of the Omega paper, with verification scripts in [`scripts/equational_theory/`](https://github.com/the-omega-institute/automath/tree/dev/theory/2026_golden_ratio_driven_scan_projection_generation_recursive_emergence/scripts/equational_theory).
